sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/AutoValueUtils.java

```diff
@@ -70,45 +72,55 @@
 /** Utilities for managing AutoValue schemas. */
 @SuppressWarnings({"rawtypes"})
 public class AutoValueUtils {
+  @SuppressWarnings("unchecked")
   public static @Nullable TypeDescriptor<?> getBaseAutoValueClass(
       TypeDescriptor<?> typeDescriptor) {
     // AutoValue extensions may be nested
     @Nullable TypeDescriptor<?> baseTypeDescriptor = typeDescriptor;
     while (baseTypeDescriptor != null
         && baseTypeDescriptor.getRawType().getName().contains("AutoValue_")) {
+      TypeDescriptor<?> finalBaseTypeDescriptor = baseTypeDescriptor;
       baseTypeDescriptor =
           Optional.ofNullable(baseTypeDescriptor.getRawType().getSuperclass())
-              .map(TypeDescriptor::of)
+              .map(superClass -> finalBaseTypeDescriptor.getSupertype((Class) superClass))
               .orElse(null);
     }
     return baseTypeDescriptor;
   }
 
-  private static TypeDescriptor<?> getAutoValueGenerated(TypeDescriptor<?> typeDescriptor) {
+  @SuppressWarnings("unchecked")
+  public static TypeDescriptor<?> getAutoValueGenerated(TypeDescriptor<?> typeDescriptor) {
     String generatedClassName = getAutoValueGeneratedName(typeDescriptor.getRawType().getName());
     try {
-      return TypeDescriptor.of(Class.forName(generatedClassName));
+      return typeDescriptor.getSubtype((Class) Class.forName(generatedClassName));
     } catch (ClassNotFoundException e) {
       throw new IllegalStateException("AutoValue generated class not found: " + generatedClassName);
     }
   }
 
-  private static @Nullable Class getAutoValueGeneratedBuilder(Class<?> clazz) {
-    Class generated;
-    try {
-      generated = Class.forName(getAutoValueGeneratedName(clazz.getName()));
-    } catch (ClassNotFoundException e) {
-      return null;
-    }
-    // Find the first generated class
-    Class base = generated;
-    while (base != null && base.getName().contains("AutoValue_")) {
-      generated = base;
-      base = base.getSuperclass();
-    }
-    String builderName = generated.getName() + "$Builder";
+  public static @Nullable TypeDescriptor<?> getAutoValueGeneratedBuilder(
+      TypeDescriptor<?> typeDescriptor) {
```

Comment:
is there a reason you remoevd the code that walks up the class hierarchy here?

Reply:
oh, probably did it by mistake, I'll bring it back